presto-main/src/main/java/com/facebook/presto/operator/aggregation/histogram/ValueStore.java

```diff
@@ -0,0 +1,187 @@
+ * <p>
+ * Note it assumes you're storing # -> Value (Type, Block, position, or the result of the ) somewhere else
+ */
+public class ValueStore
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(GroupedTypedHistogram.class).instanceSize();
+    private static final float MAX_FILL_RATIO = 0.5f;
+    private static final GroupedTypedHistogram.ProbeType PROBE_TYPE = SUM_OF_SQUARE;
+    private static final int EMPTY_BUCKET = -1;
+    private final BlockBuilder values;
+    private int rehashCount;
+
+    private int mask;
+    private int bucketCount;
+    private IntBigArray buckets;
+    private IntBigArray valueHashList;
+    private int maxfill;
+
+    @VisibleForTesting
+    public ValueStore(int expectedSize, BlockBuilder values)
+    {
+        bucketCount = computeBucketCount(expectedSize, MAX_FILL_RATIO);
+        mask = bucketCount - 1;
+        maxfill = calculateMaxFill(bucketCount, MAX_FILL_RATIO);
+        this.values = values;
+        buckets = new IntBigArray(-1);
+        buckets.ensureCapacity(bucketCount);
+        valueHashList = new IntBigArray(-1);
+        valueHashList.ensureCapacity(bucketCount);
+    }
+
+    /**
+     * cryptically named, this will add an item if not already in the system. It returns a pointer that is unique for multiple instances of the value
+     *
+     * @param type
+     * @param block
+     * @param position
+     * @param valueOnlyRawHash
+     * @return
+     */
+    public int getPosition(Type type, Block block, int position, long valueOnlyRawHash)
+    {
+        if (values.getPositionCount() >= maxfill) {
+            rehash();
+        }
+
+        long valueHash = valueOnlyRawHash;
+        int bucketId = getHashPosition(valueHash, mask);
+        int valuePointer;
+
+        // look for an empty slot or a slot containing this key
+        int probeCount = 1;
+        int originalBucketId = bucketId;
+        while (true) {
+            valuePointer = buckets.get(bucketId);
+            // empty
+            if (valuePointer == EMPTY_BUCKET) {
+                break;
+            }
+            // value match
+            if (type.equalTo(block, position, values, valuePointer)) {
+                break;
+            }
+            int probe = nextProbe(probeCount);
+            bucketId = nextBucketId(originalBucketId, mask, probe);
+            probeCount++;
+        }
+
+        if (valuePointer == EMPTY_BUCKET) {
+            valuePointer = values.getPositionCount();
+            type.appendTo(block, position, values);
+            buckets.set(bucketId, valuePointer);
+        }
+
+        checkState(probeCount < bucketCount, "we can't search more probe %s more than buckets %d", probeCount, bucketCount);
+
+        valueHashList.set(valuePointer, (int) valueHash);
+
+        return valuePointer;
+    }
+
+    private int getHashPosition(long valueHash, int mask)
```

Comment:
yea, i've changed these to getBucketId()

since all it does is translate a hash -> bucketId by & mask

computeBucketId() is fine too

either way